File: src/usart.rs
use utils::*;
use atmega328p::*;
use core::ptr::read_volatile;
use core::ptr::write_volatile;

pub const F_CPU: u32 = 16000000;
pub const BAUD: u32 = 9600;
pub const UBRR_VALUE:u32 =  ((F_CPU) + 8 * (BAUD)) / (16 * (BAUD)) -1;
pub const UBRRL_VALUE:u8 = (UBRR_VALUE & 0xff) as u8;
pub const UBRRH_VALUE:u8 = (UBRR_VALUE >> 8) as u8;

/* Public USART functions */ 
pub unsafe fn usart_init() {   
    // Set to 8 bit data size
    set_bit(UCSR0C,0x1,true); // UCSZ00 = 1
    set_bit(UCSR0C,0x2,true); // UCSZ01 = 2

    // Enable receiver and transitter
    set_bit(UCSR0B,0x3,true); // Tx enable
    set_bit(UCSR0B,0x4,true); // Rx enable

    // Set baudrate
    write_volatile(UBRR0H,UBRRH_VALUE);
    write_volatile(UBRR0L,UBRRL_VALUE);
}


pub unsafe fn readln(buff: &mut [Option<u8>]) -> &mut [Option<u8>] {
    let EOL_char: u8 = '\n' as u8;
    for i in 0..buff.len() {
        let new_byte = read_byte();
        if new_byte != EOL_char{
            buff[i] = Some(new_byte);
        } 
        else { break }
    }
    buff
}


pub unsafe fn println(msg: &str) {
    for c in msg.chars() {
        send_byte(c as u8);
    }
    send_byte('\n' as u8);
}




/* Private USART Functions */
unsafe fn tx_is_ready() -> bool {
    // Check if the UDRE0 field in UCSR0A register is 1 (ready) or 0 (not ready)
    // UDRE0 = 0x20 = 0b00100000
    if (read_volatile(UCSR0A) & UDRE0 as u8) == 0 {
         false 
    } else{ true }
}


unsafe fn rx_is_ready() -> bool {
    // Check if the RXC0 field in UCSR0A register is 1 (ready) or 0 (not ready)
    if (read_volatile(UCSR0A) & RXC0 as u8) == 0 {
        false
    } else { true }
}


pub unsafe fn send_byte(c: u8) {
    // Writes a single byte to usart buffer if its ready
    loop{ if tx_is_ready(){ break } } // do nothing until tx buffer is ready
    write_volatile(UDR0,c as u8); // when its ready, store character in tx buffer
}


unsafe fn read_byte() -> u8 {
    loop{ if rx_is_ready(){ break } }
    read_volatile(UDR0)// read and return data from buffer
}

File: baremetal/C_version/src/usart.c
#include "usart.h"

void usart_init(void) {
    UBRR0H = UBRRH_VALUE; // 0
    UBRR0L = UBRRL_VALUE; // 103
    UCSR0C = (1<<UCSZ01) | (1<<UCSZ00); /* 8-bit data */ 
    UCSR0B = (1 << TXEN0) | (1 << RXEN0); // Enable TX and RX
}

void usart_send_byte(char val) {
    while ( !( UCSR0A & (1<<UDRE0)) );
    UDR0 = val;
}

void usart_print(char *str) {
    uint16_t i;
    uint16_t len = strlen(str); 
    for (i = 0; i < len; i++){
        usart_send_byte(str[i]);
    }
}

void usart_println(char *str) {
    uint16_t i;
    uint16_t len = strlen(str); 
    for (i = 0; i < len; i++){
        usart_send_byte(str[i]);
    }
    usart_send_byte('\n');
}


char read_byte(void){
    while ( !(UCSR0A & (1<<RXC0))) {}
    return UDR0;
}

void read_str(char *buff) {
    int i = 0;
    char EOL_char = '\n';
    while(1){
        char temp_byte = read_byte();
        if (temp_byte != EOL_char) {
            buff[i] = temp_byte;
            i ++;
        } else { break; }
    }
}
